### PR TITLE
sensors: ist8310: Add Mag driver

### PR DESCRIPTION
**Pull request by PetervdPerk-NXP**
Adds support for the Isentek IST8310 3-axis magnetic sensor

### CODE REVIEW

File: drivers/sensor/ist8310/ist8310.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+#include <stdint.h>
+#include <zephyr/sys/util.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/drivers/gpio.h>
```
Comment (yperess) [inline drivers/sensor/ist8310/ist8310.h:51]: nit: duplicate includes from above and sort order

**Review by yperess — CHANGES_REQUESTED** [review]
Need to add tests for this under https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all/sensor.

@tristan-google can help as he's working on a catch all test for sensors.

**Comment by PetervdPerk-NXP** [comment]
Does just adding the following suffice, or is more needed?
```
test_i2c_ist8310@6f {
	compatible = "isentek,ist8310";
	reg = <0x6f>;
	status = "okay";
};
```

**Comment by tristan-google** [comment]
> Does just adding the following suffice, or is more needed?

Yes, that should cause the driver to build since you set up Kconfig to enable that sensor based on the device tree. I would run `scripts/twister -civ -T tests/drivers/build_all/sensor/` to ensure no compilation errors and check the output directory or logs to verify `ist8310.c` actually built. Note that without a matching emulator, this test will really only be checking that the driver builds and links (Zephyr will attempt to initialize it, but that will gracefully fail at the first I2C transaction since there is nothing on the other end --this is OK)

**Comment by PetervdPerk-NXP** [comment]
Thanks building locally using `scripts/twister -civ -T tests/drivers/build_all/sensor/` is working and CI output looks okay as well.

**Comment by bperseghetti** [comment]
@tristan-google @yperess @teburd all good now to approve/merge?

**Review by bperseghetti — APPROVED** [review]

**Comment by igalloway** [comment]
@yperess  - are you ok to approve now that the test has been added?

File: drivers/sensor/ist8310/ist8310.c [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+	}
+
+	if (ist8310_reg_read(dev, IST8310_OUTPUT_VALUE_X_L, (uint8_t *)buff, 6) < 0) {
+		LOG_ERR("failed to read mag values");
+		return -EIO;
+	}
+
+	drv_data->sample_x  = (buff[1] << 8 | buff[0]);
+	drv_data->sample_y  = (buff[3] << 8 | buff[2]);
+	drv_data->sample_z  = (buff[5] << 8 | buff[4]);
```
Comment (yperess) [inline drivers/sensor/ist8310/ist8310.c:66]: I think these are missing a call to `sys_be16_to_cpu`. Should look like:

```c
drv_data->sample_x = sys_be16_to_cpu((buff[1] << 8) | buff[0]);
```

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 3]
```diff
@@ -843,3 +843,9 @@ test_i2c_bmi08x_gyro: bmi08x@6e {
 	gyro-hz = "1000_116";
 	gyro-fs = <1000>;
 };
+
+test_i2c_ist8310@6f {
+	compatible = "isentek,ist8310";
+	reg = <0x6f>;
+	status = "okay";
+};
```
Comment (yperess) [inline tests/drivers/build_all/sensor/i2c.dtsi:851]: (I won't block on this one)

This is a good start to make sure that the driver builds at least, ideally we'd see an emulator with the driver (not yet a hard requirement in Zephyr). You can see https://github.com/zephyrproject-rtos/zephyr/tree/main/drivers/sensor/akm09918c as a sample of a similar device's emulator.

**Review by yperess — CHANGES_REQUESTED** [review]
I think there's a small bug in the way the driver was storing the values into `sample_*`, other than that just run `clang-format` and it looks ready. (would love to see an emulator with this to make sure the driver doesn't regress).

File: drivers/sensor/ist8310/ist8310.c (hunk 2 above)
Comment (PetervdPerk-NXP) [reply drivers/sensor/ist8310/ist8310.c]: Good point missed that one, made it more portable now.

File: tests/drivers/build_all/sensor/i2c.dtsi (hunk 3 above)
Comment (bperseghetti) [reply tests/drivers/build_all/sensor/i2c.dtsi:851]: Awesome thanks, we will look to add the emulation down the road with some other sensors as well.

**Review by teburd — APPROVED** [review]

**Review by yperess — APPROVED** [review]

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 4]
```diff
@@ -843,3 +843,9 @@ test_i2c_bmi08x_gyro: bmi08x@6e {
 	gyro-hz = "1000_116";
 	gyro-fs = <1000>;
 };
+
+test_i2c_ist8310@6f {
+	compatible = "isentek,ist8310";
+	reg = <0x6f>;
+	status = "okay";
```
Comment (fabiobaltieri) [inline tests/drivers/build_all/sensor/i2c.dtsi:850]: this is not needed, nodes are enabled by default

File: drivers/sensor/ist8310/ist8310.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	const struct ist8310_bus_io *bus_io;
+};
+
+struct ist8310_data {
+	struct k_sem sem;
+	int16_t sample_x, sample_y, sample_z;
+};
+
+int ist8310_reg_update_byte(const struct device *dev, uint8_t reg,
+			   uint8_t mask, uint8_t value);
```
Comment (fabiobaltieri) [inline drivers/sensor/ist8310/ist8310.h:97]: leftover? Can't find any reference to this

File: drivers/sensor/ist8310/ist8310.c [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+		LOG_ERR("failed to set single");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int ist8310_init(const struct device *dev)
+{
+	int err = 0;
```
Comment (fabiobaltieri) [inline drivers/sensor/ist8310/ist8310.c:184]: no need to initialize this one

File: drivers/sensor/CMakeLists.txt [hunk 7]
```diff
@@ -54,6 +54,7 @@ add_subdirectory_ifdef(CONFIG_IIS3DHHC iis3dhhc)
 add_subdirectory_ifdef(CONFIG_INA219 ina219)
 add_subdirectory_ifdef(CONFIG_INA23X ina23x)
 add_subdirectory_ifdef(CONFIG_INA3221 ina3221)
+add_subdirectory_ifdef(CONFIG_IST8310 ist8310)
```
Comment (fabiobaltieri) [inline drivers/sensor/CMakeLists.txt:57]: if you meant to insert this alphabetically, you almost made it :-)

Comment (fabiobaltieri) [reply drivers/sensor/CMakeLists.txt:57]: if I meant to make this review before merge, I almost made it :-D

File: drivers/sensor/ist8310/ist8310.c [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	}
+
+	if (ist8310_reg_read(dev, IST8310_OUTPUT_VALUE_X_L, (uint8_t *)buff, 6) < 0) {
+		LOG_ERR("failed to read mag values");
+		return -EIO;
+	}
+
+	drv_data->sample_x = (sys_le16_to_cpu(*(uint16_t *)&buff[0]));
+	drv_data->sample_y = (sys_le16_to_cpu(*(uint16_t *)&buff[2]));
+	drv_data->sample_z = (sys_le16_to_cpu(*(uint16_t *)&buff[4]));
```
Comment (fabiobaltieri) [inline drivers/sensor/ist8310/ist8310.c:64]: you could drop the outer parenthesis here

**Comment by bperseghetti** [comment]
@fabiobaltieri I can open a different PR with those changes, or when we add the emulation for sensors with @tristan-google and @yperess work we could make those changes then.

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri I can open a different PR with those changes, or when we add the emulation for sensors with @tristan-google and @yperess work we could make those changes then.

Either works, thanks.

**Pull request merged**